xdsl/dialects/builtin.py

```diff
@@ -40,12 +40,42 @@ def verify(self, attr: Attribute) -> None:
 
 
 _ArrayAttrT = TypeVar("_ArrayAttrT", bound=Attribute, covariant=True)
+_ArrayAttrInvT = TypeVar("_ArrayAttrInvT", bound=Attribute)
 
 
 @irdl_attr_definition
 class ArrayAttr(GenericData[List[_ArrayAttrT]]):
     name: str = "array"
 
+    @overload
+    @staticmethod
+    def is_array_of(value: Any) -> TypeGuard[ArrayAttr[Attribute]]:
+        ...
+
+    @overload
+    @staticmethod
+    def is_array_of(
+            value: Any, type: type[_ArrayAttrInvT]
+    ) -> TypeGuard[ArrayAttr[_ArrayAttrInvT]]:
+        ...
+
+    @staticmethod
+    def is_array_of(
+        value: Any,
+        type: type[_ArrayAttrInvT] = Attribute
+    ) -> TypeGuard[ArrayAttr[_ArrayAttrInvT]]:
+        # Check if the base type is correct
+        if not isinstance(value, ArrayAttr):
+            return False
+        value = cast(ArrayAttr[Attribute], value)
+
+        # ArrayAttr only contains attributes
+        if type == Attribute:
+            return True
```

Comment:
I don't think it's wrong, but I find it confusing. Given that the test would true either way, is that an optimization we want?
As said in the overall comment, I think for this method I would prefer two methods names instead of an overload, which would read clearer for this one too (the ArrayAttr[T] could use the ArrayAttr[Attribute])

Reply:
Yes, I separated it in two functions, so now it's not needed anymore!